### PR TITLE
Fixes SendAsyncFile flaky failures

### PR DESCRIPTION
**Pull request by esdrubal**
Stack traces of blocked threads are now displayed on the exception. This help us determining which system calls are not aborted properly.

Before closing a socket we try to interrupt and wait for all socket
pending system calls to return. For that we set the socket as non
blocking which makes the system calls to return with EAGAIN\EWOULDBLOCK.

This issue was caught by System.Net.Sockets.SocketTest.SendAsyncFile
that was intermittently failing, we were only able to reproduce it while
running multiple test suites in parallel. We now can say that it only failed when
the the socket was closed and the sendfile was waiting for the OS to
allocate enough resources to send the file.

This was fixed by changing wapi_sendfile to return on EAGAIN.

Thanks to @akoeplinger  suggestions I was able to login into Jenkins bots and reproduce the flaky failure, I was also able to confirm that with the fix the issue no longer occurred while looping two instance of System tests in parallel for more than one hour.

### CODE REVIEW

File: mcs/class/System/System.Net.Sockets/SafeSocketHandle.cs [hunk 1]
```diff
@@ -48,8 +55,17 @@ protected override bool ReleaseHandle ()
 					int abort_attempts = 0;
 					while (blocking_threads.Count > 0) {
 						if (abort_attempts++ >= ABORT_RETRIES) {
-							if (THROW_ON_ABORT_RETRIES)
-								throw new Exception ("Could not abort registered blocking threads before closing socket.");
+							if (THROW_ON_ABORT_RETRIES) {
+								StringBuilder sb = new StringBuilder ();
+								sb.AppendLine ("Could not abort registered blocking threads before closing socket.");
+								foreach (var thread in blocking_threads) {
+									sb.AppendLine ("Thread StackTrace:");
```
Comment (akoeplinger) [inline mcs/class/System/System.Net.Sockets/SafeSocketHandle.cs]: Adding Thread.Name/Thread.ManagedThreadId here would probably make it even better.

**Comment by akoeplinger** [comment]
LGTM, thank you so much for tracking this down!

**Comment by alexrp** [comment]
Is this likely to fix the `SendAsync` failure on Android as well? Specifically this:

```
MonoTests.System.Net.Sockets.SocketAsyncTest.SendAsync=System.TimeoutException : The operation has timed out.
		  at MonoTests.System.Net.Sockets.SocketAsyncTest.SendAsync () [0x00141] in /Users/builder/data/lanes/1196/89794de1/source/monodroid/tests/bcl-test/src/System/Test/System.Net.Sockets/SocketAsyncTest.cs:134 
		  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
		  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /Users/builder/data/lanes/1196/6a8eb978/source/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:309 
```

**Comment by esdrubal** [comment]
@alexrp I don't think this will affect the SendAsync failure as this only changes the unrelated native function wapi_sendfile. Without debugging it is hard to tell why it is failing.

**Comment by esdrubal** [comment]
The bot failures do not look related, merging this.

**Comment by akoeplinger** [comment]
@esdrubal can you please put it into 4.5.1 too?

**Pull request merged**